### ISSUE TITLE
driver: undocumented max_velocity parameter causes plan execution to fail

### ISSUE THREAD

**Issue by gavanderhoorn** [ros-industrial/universal_robot#124]
As reported on ROS Answers in [path planning exceeding velocity limits of ur5](http://answers.ros.org/question/194667/path-planning-exceeding-velocity-limits-of-ur5/). Planning with MoveIt often results in:

```
[ERROR] [WallTime: 1412925263.358047] Received a goal with velocities that are higher than 2.000000
```

The private parameter `max_velocity` is currently undocumented (as are all other parameters).

The wiki page for `ur_driver` should probably be extended with an overview of the defined parameters and their values.

**Comment by fmessmer** [comment]
Well MoveIt! plans with the velocity_limits given in [joint_limits.yaml](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur5_moveit_config/config/joint_limits.yaml)
And those values are retrieved from the velocity limit tags in the URDF description of the joints (see e.g. [here](https://github.com/ros-industrial/universal_robot/blob/hydro-devel/ur_description/urdf/ur5.urdf.xacro#L105)).

With a `max_velocity` parameter in the driver having a value less than the MoveIt! velocity limits, this would most likely cause most trajectories to fail.

I thus would increase the default `max_velocity` parameter in the driver to (the maximum) of the MoveIt velocity limits (or have velocity limits per joint in the driver?) in order to allow executing the planned trajectories.

Everyone who wants to move at a slower speed would need to reduce both `max_velocity` parameter in the driver and the MoveIt! velocity limits accordingly.
In order to facilitate this, we could add the private parameter to the ur_common.launch file as suggested by @gavanderhoorn 

Any comments?

**Comment by shaun-edwards** [comment]
I have safely increased the max velocity parameter on my system.  I believe the outlined approach is the right one to take.  It will lead to "safety stops" by the universal because the current driver isn't capable of moving much faster than the `max_velocity` reliably.

**Comment by gavanderhoorn** [comment]
On 30-10-2014 4:19, Shaun Edwards wrote:

> I have safely increased the max velocity parameter on my system.  I believe the outlined approach is the right one to take.  It will lead to "safety stops" by the universal because the current driver isn't capable of moving much faster than the `max_velocity` reliably.  

I think the reason I reported this was not so much the technical issue
of having a parameter that limits the max joint speed the driver is
'comfortable' allowing in a trajectory, but more the fact that that
parameter is completely undocumented, and (worse) has a default that is
so low that normal usage of the released packages will result in errors
that complain about a 'limit' being exceeded. As can be seen in the ROS
Answers post I linked, users understand this to be a limit of the
actual, physical robot, and start looking for ways to scale down the
velocities MoveIt uses for planning.

I agree that the default should probably just be put at something like
100 (so no limit by default), but in any case this (and other)
parameters of the Python node should be documented on the ROS wiki.

**Comment by tnl-woi** [comment]
I agree it should be well documented which parameters there are and what they do, so that users know how they can control the speed of the robot.

Furthermore, the default value should not cause any problems, so the current value is too low. I agree with @ipa-fxm that it should be at least the maximum of the MoveIt velocity limits, or a little bit larger to allow some margin. I would certainly not set it higher than 10, because that's already way over the physical limit of the robot.
This should indeed be a parameter to the launch file, making it configurable using the command-line.

I don't think it's necessary to have a limit for each joint, because I think this parameter is more of a 'sanity check' or temporary debug option and I would not add too much code/complexity to the driver in order to facilite that. One global max is good enough.

However, it would be very nice if the values in joint_limits.yaml could be updated in the same way (using a command-line parameter) to actually generate slower trajectories instead of failing on fast trajectories. So if that's possible, I would certainly like that. I guess this would also be a very good work-around for the known safety stop issues.

**Comment by gavanderhoorn** [comment]
@tnl-woi wrote:

> I agree with @ipa-fxm that it should be at least the maximum of the MoveIt velocity limits, or a little bit larger to allow some margin. I would certainly not set it higher than 10, because that's already way over the physical limit of the robot.

If the user does not set a maximum herself, and MoveIt already always uses the max joint velocities, I don't really see a need for `max_velocity` being "a little bit larger" than the actual limits. The actual value doesn't matter, but it's not necessarily related to the joint limits.

But, to stay on-topic, this issue just asks three things:
1. increase the default value of the `max_velocity` parameter
2. document it on the wiki (using the regular format for documenting node parameters) and make clear it has _nothing_ to do with limitations of the actual robot
3. log whatever value `max_velocity` actually has to the console, on starting the driver

@ipa-fxm wrote:

> Everyone who wants to move at a slower speed would need to reduce both `max_velocity` parameter in the driver and the MoveIt! velocity limits accordingly.

Let's just make it clear in the documentation that `max_velocity` is for debugging purposes only, and that the default value is so high that it does not interfere with normal operations. Anyone who wants to use it as a makeshift safety-like feature may do so, but should understand what the implications are (ie: other configuration files will need to be updated).

**Comment by gavanderhoorn** [comment]
I've just updated the [ur_driver](http://wiki.ros.org/ur_driver) wiki page. At least the documentation part is taken care of.

I've documented the current `-devel` branches of the driver, so some information may not match the current debs.

**Comment by fmessmer** [comment]
Thanks for doing the documentation part @gavanderhoorn 
I just did the rest in #136 

> However, it would be very nice if the values in joint_limits.yaml could be updated in the same way (using a command-line parameter) to actually generate slower trajectories instead of failing on fast trajectories. So if that's possible, I would certainly like that. I guess this would also be a very good work-around for the known safety stop issues.

@tnl-woi 
I guess that's beyond the scope of the universal_robot repository...it's just the nature of yaml files to be used for configuration...setting all those parameters from joint_limits.yaml from a launch file would be messy...

**Comment by abubeck** [comment]
Fixed in #136 .
